### PR TITLE
feat: (Platform) bring Form Grid styles for Platform Forms

### PR DESCRIPTION
**Pull request by kavya-b**
#### Please provide a link to the associated issue.
#3806 
#### Please provide a brief summary of this pull request.
Currently, Platform Forms uses 3rd party library flexboxgrid to achieve responsive layouts. This PR replaces it with Form Grid from Fundamental Styles.
Group Header styles are not implemented, therefore Group Header integration will happen separately at a later point.

BREAKING CHANGE:
A new input field `columnLayoutType` introduced that takes the column layout in the format `XLn-Ln-Mn-Sn` where n is the number of columns and can be different for each size.
For eg: XL2-L2-M2-S1 would create 2-column layouts for XL, L, and M sizes and single-column layout for S size.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] update `README.md`
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

### CODE REVIEW

**Comment by KevinOkamoto** [comment]
@kavya-b were there requirements or tech docs for this code change? I ask because, @fkolar and I had been discussing the architecture for this, and it involved adding `column` attribute to the form field as well as `groupLabel` or something to that effect. I don't think we formally wrote down our thoughts in a tech doc, or if we even came to a final decision on the architecture. If there is a tech doc, can you add the link here.

**Comment by kavya-b** [comment]
> @kavya-b were there requirements or tech docs for this code change? I ask because, @fkolar and I had been discussing the architecture for this, and it involved adding `column` attribute to the form field as well as `groupLabel` or something to that effect. I don't think we formally wrote down our thoughts in a tech doc, or if we even came to a final decision on the architecture. If there is a tech doc, can you add the link here.

This PR only brings the existing styles for Form Grid from Fundamental Styles and removes the `flexboxgrid` library being used currently in ngx/platform. Work on form groups with column and group header/label attributes will be taken up separately(probably by a Valor member) as it definitely needs more discussion. 
Fundamental Styles also does not have any style implementation for Group Headers as of now(I am working on it in Styles).
cc: @manjunathanagaraj

**Review by KevinOkamoto — APPROVED** [review]
Looks good

File: libs/platform/src/lib/components/form/form-group/form-group.component.ts [hunk 1]
```diff
@@ -458,6 +474,36 @@ export class FormGroupComponent implements FormGroupContainer, OnInit, AfterCont
         }
         return;
     }
+
+    /**
+     * @hidden
+     * if `columnLayoutType` is given, set those column layouts appropriately
+     */
+    private _setUserSpecifiedLayout(groupField: GroupField): void {
+        const layoutString = this.columnLayoutType.split('-');
```
Comment (mykolavorobiov) [inline libs/platform/src/lib/components/form/form-group/form-group.component.ts:483]: You can use destructuring assignment, `const [x,l,m]= this.columnLayoutType.split('-');`
and more readable
`parseInt(x.slice(2, x.length), 10);`

File: libs/platform/src/lib/components/form/form-group/form-group.component.ts [hunk 2]
```diff
@@ -458,6 +474,36 @@ export class FormGroupComponent implements FormGroupContainer, OnInit, AfterCont
...
+            throw new Error('Columns cannot be more than 12');
+        } else {
+            const lgColumns = 12 / lgColumnsNumber;
+            const mdColumns = 12 / mdColumnsNumber;
+            const xlColumns = 12 / xlColumnsNumber;
+
+            // for `lg` single-column layout, Styles does not use any class, and providing `fd-col-lg--12` has unintended side-effects
+            // therefore, we remove the lg class for single-column layout
+            if (lgColumns === 12) {
+                groupField.styleClass = `fd-col-xl--` + xlColumns + ` fd-col-md--` + mdColumns;
```
Comment (mykolavorobiov) [inline libs/platform/src/lib/components/form/form-group/form-group.component.ts:500]: Please use one styles for using variable in string
```
`fd-col-xl--${xlColumns} fd-col-md--${mdColumns}`

File: libs/platform/src/lib/components/form/form-group/form-group.component.ts [hunk 3]
```diff
@@ -458,6 +474,36 @@ export class FormGroupComponent implements FormGroupContainer, OnInit, AfterCont
...
+            } else {
+                groupField.styleClass =
+                    `fd-col-xl--` + xlColumns + ` fd-col-md--` + mdColumns + ` fd-col-lg--` + lgColumns;
```
Comment (mykolavorobiov) [inline libs/platform/src/lib/components/form/form-group/form-group.component.ts:503]: Same as above

**Review by mykolavorobiov — CHANGES_REQUESTED** [review]

File: libs/platform/src/lib/components/form/form-group/form-group.component.ts (hunk 2 above)
Comment (kavya-b) [reply libs/platform/src/lib/components/form/form-group/form-group.component.ts]: fixed as suggested

File: libs/platform/src/lib/components/form/form-group/form-group.component.ts (hunk 1 above)
Comment (kavya-b) [reply libs/platform/src/lib/components/form/form-group/form-group.component.ts]: fixed as suggested, thanks for the comment!

**Review by mykolavorobiov — APPROVED** [review]

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-complex-example.component.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    selector: 'fdp-platform-form-container-complex-example',
+    templateUrl: './platform-form-container-complex-example.component.html',
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class PlatformFormContainerComplexExampleComponent {
+    form: FormGroup;
+    seasons: string[] = ['Winter', 'Spring', 'Summer', 'Autumn'];
+    qty = 10;
+
+    constructor() {
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-complex-example.component.ts:14]: plz remove

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-complex-example.component.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
+import { ChangeDetectionStrategy, Component } from '@angular/core';
+import { FormGroup } from '@angular/forms';
+
+@Component({
+    selector: 'fdp-platform-form-container-complex-example',
+    templateUrl: './platform-form-container-complex-example.component.html',
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class PlatformFormContainerComplexExampleComponent {
+    form: FormGroup;
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-complex-example.component.ts:10]: `form: FormGroup = new FormGroup({})`

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-not-recommended-example.component.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    templateUrl: './platform-form-container-not-recommended-example.component.html',
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class PlatformFormContainerNotRecommendedExampleComponent {
+    form: FormGroup;
+    form1: FormGroup;
+
+    seasons: string[] = ['Winter', 'Spring', 'Summer', 'Autumn'];
+
+    constructor() {
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-not-recommended-example.component.ts:15]: the same comments

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-possible-example.component.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+    templateUrl: './platform-form-container-possible-example.component.html',
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class PlatformFormContainerPossibleExampleComponent {
+    form: FormGroup;
+    form1: FormGroup;
+
+    seasons: string[] = ['Winter', 'Spring', 'Summer', 'Autumn'];
+
+    constructor() {
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-possible-example.component.ts:15]: the same

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-recommended-example.component.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class PlatformFormContainerRecommendedExampleComponent {
+    form: FormGroup;
+    form1: FormGroup;
+    form2: FormGroup;
+
+    seasons: string[] = ['Winter', 'Spring', 'Summer', 'Autumn'];
+
+    constructor() {
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-recommended-example.component.ts:16]: the same

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-header/platform-form-container-header.component.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { Component, OnInit } from '@angular/core';
+
+@Component({
+    selector: 'app-form-container-header',
+    templateUrl: './platform-form-container-header.component.html',
+    styleUrls: ['./platform-form-container-header.component.scss']
+})
+export class PlatformFormContainerHeaderComponent implements OnInit {
+    constructor() {}
```
Comment (Betrozov) [inline apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-header/platform-form-container-header.component.ts:9]: plz remove empty blocks

File: package.json [hunk 10]
```diff
@@ -101,7 +101,6 @@
     "@types/node": "14.11.7",
     "classlist.js": "1.1.20150312",
     "core-js": "3.7.0",
-    "flexboxgrid": "6.3.1",
```
Comment (Betrozov) [inline package.json:104]: should removing make by another PR?

**Review by Betrozov — CHANGES_REQUESTED** [review]

File: apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-complex-example.component.ts (hunk 5 above)
Comment (kavya-b) [reply apps/docs/src/app/platform/component-docs/platform-forms/platform-form-container/platform-form-container-examples/platform-form-container-complex-example.component.ts]: changed as suggested and removed the constructors, please check. Thanks!

File: package.json (hunk 10 above)
Comment (kavya-b) [reply package.json:99]: flexboxgrid was only used by platform forms, and this PR gives a replacement for it. Does it really need another PR just to remove this lib usage?

**Review by Betrozov — APPROVED** [review]

**Review by fkolar — CHANGES_REQUESTED** [review]
Maybe I missed the tech spec. Do we have any?  Just bring this in is really great move, but we havent talked about signature. 

I see you are still using zone, which in this case does not make much sense, we should be referring.  to columns or rows. 


When this new Grid layout zone should be obsolete and we should replace them with something else.  Column=1,2,3,4 

And we need to check that column number matches with the layout. When we discussed this with Kevin we were talking about something close to this:

What do you think ? 


```
<fdp-form-group layoutPattern="XL3-L3-M2-S1" >
  <fdp-form-field label="First Name" column="1" rank="1" />
  <fdp-form-field label="Last Name" column="1" rank="20" />
  <fdp-form-field label="Birthday " column="1" rank="100" />

  <fdp-form-field label="Department" column="2" rank="1" />
  <fdp-form-field label="Supervisr" column="2" rank="20" />

  <fdp-form-field label="Mobile" column="3" rank="1" />
  <fdp-form-field label="twitter" column="3" rank="120" />    
</fdp-form-group>

```

**Comment by kavya-b** [comment]
> Maybe I missed the tech spec. Do we have any? Just bring this in is really great move, but we havent talked about signature.
> 
> I see you are still using zone, which in this case does not make much sense, we should be referring. to columns or rows.
> 
> When this new Grid layout zone should be obsolete and we should replace them with something else. Column=1,2,3,4
> 
> And we need to check that column number matches with the layout. When we discussed this with Kevin we were talking about something close to this:
> 
> What do you think ?
> 
> ```
> <fdp-form-group layoutPattern="XL3-L3-M2-S1" >
>   <fdp-form-field label="First Name" column="1" rank="1" />
>   <fdp-form-field label="Last Name" column="1" rank="20" />
>   <fdp-form-field label="Birthday " column="1" rank="100" />
> 
>   <fdp-form-field label="Department" column="2" rank="1" />
>   <fdp-form-field label="Supervisr" column="2" rank="20" />
> 
>   <fdp-form-field label="Mobile" column="3" rank="1" />
>   <fdp-form-field label="twitter" column="3" rank="120" />    
> </fdp-form-group>
> ```

I understand Frank that this PR does not complete the full picture for Forms. Apart from the `columns` input you mentioned, we would also need Group Headers, for which styles were not available at the time of creating this PR. I worked on adding Group Header Fundamental Styles in PR #[1881](https://github.com/SAP/fundamental-styles/pull/1881), but integrating this into Platform Forms and `columns` feature are things that will be taken up in the next phase, mostly worked on by Evgeny, who will also be presenting the tech spec, where more discussion can happen.

**Comment by fkolar** [comment]
> > Maybe I missed the tech spec. Do we have any? Just bring this in is really great move, but we havent talked about signature.
> > I see you are still using zone, which in this case does not make much sense, we should be referring. to columns or rows.
> > When this new Grid layout zone should be obsolete and we should replace them with something else. Column=1,2,3,4
> > And we need to check that column number matches with the layout. When we discussed this with Kevin we were talking about something close to this:
> > What do you think ?
> > ```
> > <fdp-form-group layoutPattern="XL3-L3-M2-S1" >
> >   <fdp-form-field label="First Name" column="1" rank="1" />
> >   <fdp-form-field label="Last Name" column="1" rank="20" />
> >   <fdp-form-field label="Birthday " column="1" rank="100" />
> > 
> >   <fdp-form-field label="Department" column="2" rank="1" />
> >   <fdp-form-field label="Supervisr" column="2" rank="20" />
> > 
> >   <fdp-form-field label="Mobile" column="3" rank="1" />
> >   <fdp-form-field label="twitter" column="3" rank="120" />    
> > </fdp-form-group>
> > ```
> 
> I understand Frank that this PR does not complete the full picture for Forms. Apart from the `columns` input you mentioned, we would also need Group Headers, for which styles were not available at the time of creating this PR. I worked on adding Group Header Fundamental Styles in PR #[1881](https://github.com/SAP/fundamental-styles/pull/1881), but integrating this into Platform Forms and `columns` feature are things that will be taken up in the next phase, mostly worked on by Evgeny, who will also be presenting the tech spec, where more discussion can happen.
> 
> Kevin brought the [same point](https://github.com/SAP/fundamental-ngx/pull/3807#issuecomment-726247658) up in this PR few days back. :)

We can definitelly bring headers latter on. I agree. But we shoudl not start on the devs, unless we agree on the signature. Let's park this. 

Collect all the possible pages, from SAP, 1P, Our designers to see if are layout is able to address them and how, then we will be albe to craft component signature towards something that could be usable and coudl work.

**Comment by kavya-b** [comment]
> > > Maybe I missed the tech spec. Do we have any? Just bring this in is really great move, but we havent talked about signature.
> > > I see you are still using zone, which in this case does not make much sense, we should be referring. to columns or rows.
> > > When this new Grid layout zone should be obsolete and we should replace them with something else. Column=1,2,3,4
> > > And we need to check that column number matches with the layout. When we discussed this with Kevin we were talking about something close to this:
> > > What do you think ?
> > > ```
> > > <fdp-form-group layoutPattern="XL3-L3-M2-S1" >
> > >   <fdp-form-field label="First Name" column="1" rank="1" />
> > >   <fdp-form-field label="Last Name" column="1" rank="20" />
> > >   <fdp-form-field label="Birthday " column="1" rank="100" />
> > > 
> > >   <fdp-form-field label="Department" column="2" rank="1" />
> > >   <fdp-form-field label="Supervisr" column="2" rank="20" />
> > > 
> > >   <fdp-form-field label="Mobile" column="3" rank="1" />
> > >   <fdp-form-field label="twitter" column="3" rank="120" />    
> > > </fdp-form-group>
> > > ```
> > 
> > 
> > I understand Frank that this PR does not complete the full picture for Forms. Apart from the `columns` input you mentioned, we would also need Group Headers, for which styles were not available at the time of creating this PR. I worked on adding Group Header Fundamental Styles in PR #[1881](https://github.com/SAP/fundamental-styles/pull/1881), but integrating this into Platform Forms and `columns` feature are things that will be taken up in the next phase, mostly worked on by Evgeny, who will also be presenting the tech spec, where more discussion can happen.
> > Kevin brought the [same point](https://github.com/SAP/fundamental-ngx/pull/3807#issuecomment-726247658) up in this PR few days back. :)
> 
> We can definitelly bring headers latter on. I agree. But we shoudl not start on the devs, unless we agree on the signature. Let's park this.
> 
> Collect all the possible pages, from SAP, 1P, Our designers to see if are layout is able to address them and how, then we will be albe to craft component signature towards something that could be usable and coudl work.

Are you suggesting we push this PR back to WIP and once the tech spec is discussed and implementation approach is clear, all those changes can also be worked on in this same PR?

**Comment by kavya-b** [comment]
Closing this PR for now as more work will be coming into it.
